### PR TITLE
Add text support to the Trainer's TensorBoard integration

### PR DESCRIPTION
**Pull request by JacobLinCool**
# What does this PR do?

Add text support to `TensorBoardCallback` so users can see the decoded text generated in `compute_metrics` in TensorBoard. Sometimes the string is very long, so a `max_str_len` (defaults to 100) property has also been added to `ProgressCallback` to prevent polluting the progress log.

Can be used to checkout transcription when fine-tuning ASR models:
截圖 2024-10-26 凌晨2 49 07

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- trainer: @muellerzr and @SunMarc

### CODE REVIEW

File: src/transformers/trainer_callback.py [hunk 1]
```diff
@@ -631,7 +632,10 @@ def on_log(self, args, state, control, logs=None, **kwargs):
             # but avoid doing any value pickling.
             shallow_logs = {}
             for k, v in logs.items():
-                shallow_logs[k] = v
+                if isinstance(v, str) and len(v) > self.max_str_len:
+                    shallow_logs[k] = f"[String too long to display, length: {len(v)}]"
```
Comment (SunMarc) [inline src/transformers/trainer_callback.py:636]: Maybe you can even tell them to modify max_str_len in the callback if they want to increase the length

**Review by SunMarc — APPROVED** [review]
Thanks a lot ! LGMT

**Review by LysandreJik — APPROVED** [review]
LGTM!

File: src/transformers/trainer_callback.py (hunk 1 above)
Comment (JacobLinCool) [reply src/transformers/trainer_callback.py]: I've added some docstrings and included a hint in the message.

Comment (JacobLinCool) [reply src/transformers/trainer_callback.py]: Hi @SunMarc, just checking in to see if there's any update on this PR. Please let me know if any additional changes are needed. Thanks!

**Comment by LysandreJik** [comment]
Feel free to merge when good @SunMarc

File: src/transformers/trainer_callback.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/trainer_callback.py]: Nice, thanks for your work !

**Review by SunMarc — APPROVED** [review]
Awesome ! Thanks for integrating my suggestion.

**Pull request merged**